File: Pieces/pawn.py
from Pieces.piece import Piece
from enums import Colour
from coords import Coords

from Pieces.moveCandidate import MoveCandidate
from Pieces.rook import Rook
from Pieces.queen import Queen
from Pieces.bishop import Bishop
from Pieces.knight import Knight

class Pawn(Piece):
    def __init__(self, colour: Colour) -> None:
        super().__init__(colour)

    
    #TODO Change to __str__ or __repr__
    def get_representation(self) -> str:
        return 'p' if self.colour == Colour.BLACK else 'P'

    def get_candidate_moves(self, coords: Coords):
        moves = []

        direction = 1 if self.colour == Colour.WHITE else -1

        # checking forward moves
        moves_forward = 2 if \
            (coords.rank == 2 and self.colour == Colour.WHITE) \
            or (coords.rank == 7 and self.colour == Colour.BLACK) else 1

        move = MoveCandidate(
            direction,
            0,
            moves_forward,
            False,
            False
        )

        moves.append(move)

        # check captures
        move = MoveCandidate(
            direction,
            -1,
            1,
            True,
            True,
        )
        moves.append(move)

        move = MoveCandidate(
            direction,
            1,
            1,
            True,
            True,
        )
        moves.append(move)

        return moves

    

File: rust_chess/src/pieces/pawn.rs
use crate::enums::Colour;
use crate::coords::Coords;
use crate::moves::move_ray::MoveRay;
use crate::pieces::Piece;

#[derive(Debug, Clone)]
pub struct Pawn {
    pub colour: Colour,
}

impl Pawn {
    pub fn new(colour: Colour) -> Self {
        Self { colour }
    }
}

impl Piece for Pawn {
    fn colour(&self) -> Colour {
        self.colour
    }

    fn get_representation(&self) -> char {
        match self.colour {
            Colour::White => 'P',
            Colour::Black => 'p',
        }
    }

    fn get_destination_coords(&self, coords: Coords) -> Vec<MoveRay> {
        let mut moves = Vec::new();
        let direction: i8 = if self.colour == Colour::White { 1 } else { -1 };

        // Forward moves
        let moves_forward = if (coords.rank == 2 && self.colour == Colour::White)
            || (coords.rank == 7 && self.colour == Colour::Black)
        {
            2
        } else {
            1
        };

        moves.push(MoveRay {
            rank_diff: direction,
            file_diff: 0,
            dist: moves_forward,
            capture_allowed: false,
            capture_forced: false,
        });

        // Capture left
        moves.push(MoveRay {
            rank_diff: direction,
            file_diff: -1,
            dist: 1,
            capture_allowed: true,
            capture_forced: true,
        });

        // Capture right
        moves.push(MoveRay {
            rank_diff: direction,
            file_diff: 1,
            dist: 1,
            capture_allowed: true,
            capture_forced: true,
        });

        moves
    }
}